### ISSUE TITLE
Support const_fn

### ISSUE THREAD

**Issue by rasendubi** [alexcrichton/futures-rs#463]
I would really like if futures have `const fn` whenever that possible, for that allows me to create `static` futures and initialize them without allocating heap or stack memory, which is nice in embedded environment. (yeah, I'm running futures on top of a bare metal OS, interrupts, and a simple reactor.)

I don't mind implementing this request myself, but I have an issue I can't solve.

As `const_fn` is a nightly-only, adding them will break the stable build. I have considered using Cargo features so that I can enable `const_fn` feature gate by request, but that seem to require implementing each function twice: one with `const` when `cfg(feature = "with-const-fn")` and another when the feature is disabled.

I could solve that with a macro, but that's not nice.

**Comment by alexcrichton** [comment]
Oh interesting! I didn't know that this was a use case for futures. Mind expanding a bit on sticking futures in statics? (I'm mostly just curious)

**Comment by rasendubi** [comment]
Sure. I am building a small experimental embedded system ([bkernel](https://github.com/rasendubi/bkernel)) and I use futures as a main source of multi-tasking. Most of the data come from and go to the hardware interfaces, and so far that nicely fits with Streams and Sinks traits. ([Here](https://github.com/rasendubi/bkernel/blob/4cc6600d44b64ef1563f57fa3223e9878155b6b8/src/log.rs) is an example of adaptors for USART.)

[My reactor](https://github.com/rasendubi/bkernel/blob/4cc6600d44b64ef1563f57fa3223e987814cc6600d44b64ef1563f57fa3223e9878155b6b8/breactor/src/lib.rs) allows a maximum of 32 active futures (tasks) and uses a bitmask to designate readiness. (Task number is also used as a priority.) The readiness is set from interrupts.

The thing is, most of my [current tasks](https://github.com/rasendubi/bkernel/blob/4cc6600d44b64ef1563f57fa3223e987814cc6600d44b64ef1563f57fa3223e9878155b6b8/src/lib.rs#L78-L94) merely process data coming from a stream enqueuing responses into a sink. Practically, they never end, and I don't spawn new tasks dynamically. (I am currently not using heap memory at all.)

So the main (and only) thread allocates these futures on the stack and runs reactor for completion—i.e. forever (putting the processor to sleep when no future is ready) ([src/lib:97-119](https://github.com/rasendubi/bkernel/blob/4cc6600d44b64ef1563f57fa3223e987814cc6600d44b64ef1563f57fa3223e9878155b6b8/src/lib.rs#L97-L119)).

It would be nice to stick futures into statics, so I could live without transmuting lifetimes and also save little of stack.

**Comment by alexcrichton** [comment]
Oh nice! 

We could possibly enable this with a feature on the crate, but that may mean a lot of duplication (unsure how this is done elsewhere). In general though exporting a `const fn` is a pretty big commitment because it means that we can *never* switch to something requiring, for example, an `if` or allocation or anything like that. I don't think we'd want to do so but it's tough to say for sure that we'd never want to do so.

**Comment by alexcrichton** [comment]
Note that if you don't need a lot of combinators the implementations are quite small and may be inlineable locally as something like `const_map`, for example?

**Comment by rasendubi** [comment]
Sure. It's just would be nice if the library provides them out of the box :)

**Comment by rasendubi** [comment]
Ok, faced another issue with static futures. Even though I have created a const fn version of lazy future, it is extremely painful to instantiate it, as statics require spelling out the full type (which gets unmanageable pretty fast with combinators).

e.g., ``futures::future::MapErr<futures::future::Map<futures::future::AndThen<start_send_all_string::StartSendAllString<'_, &mut log::IoBuffer>, impl futures::Future, [closure@src/lib.rs:123:16: 123:62 stdin:_]>, [closure@src/lib.rs:124:14: 124:20]>, [closure@src/lib.rs:125:18: 125:24]>`
``

Note that it is not even possible to spell due to closures. (Though, `closure_to_fn_coercion` might help.)

I've been using `impl Trait` to avoid that in function types, but that didn't quite work. Even though this allows writing a function, I am unable to store its result in a struct field (as I can't spell a concrete type). 

So basically I am forced to implement my own future state machines manually without using combinators (almost):
- https://github.com/rasendubi/bkernel/commit/da583775fd68cf02f54e72599547eeefffea6b7b
- https://github.com/rasendubi/bkernel/commit/6cd78ab0b8a56b1593a8c406f126adf68f83d72e

... dreaming of `typeof`...

**Comment by alexcrichton** [comment]
I'm going to close this for now as `const fn` isn't currently stable, but when it's stable we can consider making use of it!
